source/Messages.cpp

```diff
@@ -33,6 +39,9 @@ void Messages::Add(const string &message, Importance importance)
 {
 	lock_guard<mutex> lock(incomingMutex);
 	incoming.emplace_back(message, importance);
+	log.emplace(log.begin(), message, importance);
```

Comment:
This isn't ideal because the vector has to more every element over every single time a message is added. `std::queue` is probably a better fit (or similar).

Reply:
`std::queue` isn't actually a container. It's an adaptor that you pass a container to and it will expose queue relevant methods for that container.
We'll need to implement our own queue.
https://github.com/warp-core/endless-sky/tree/Queue
This is a circular queue using a vector.

Reply:
>  We'll need to implement our own queue.

Huh? Why not use `std::queue`? Yeah its an abstraction over a `std::deque` by default, but that's fine?

Reply:
An `std::deque` while not moving the data around, is still going to need to allocate and deallocate blocks at the beginning and end as stuff gets added. It also has slightly worse random access time than a vector. While this isn't going to be a huge deal given that iteration occurs while the main view is paused, it's something.
I don't think not needing to allocate the memory all at once and not needing that memory to be contiguous are really very big advantages in this case given that we're looking at string int pairs, even 10000 of them aren't going to take up that much space.

Reply:
>  is still going to need to allocate and deallocate blocks at the beginning and end as stuff gets added.

just like a vector no?

> It also has slightly worse random access time than a vector. While this isn't going to be a huge deal given that iteration occurs while the main view is paused, it's something.

yeah but like you said, it's UI so not really a big deal. It's not like this is slow, iterating a std::map is even worse but we do that like everywhere.

> I don't think not needing to allocate the memory all at once and not needing that memory to be contiguous are really very big advantages in this case given that we're looking at string int pairs, even 10000 of them aren't going to take up that much space.

Yeah a circular buffer would indeed be better (boost has a nice one that fits exactly this use case too), but I think std::queue is good enough for our purposes and avoids having to write a circular buffer implementation.

Reply:
So, what's our best option?

Reply:
I agree with quyykk in this case since performance in UI is no big deal.